core/src/test/java/org/helsinki/paytrail/request/refunds/PaytrailRefundCreateTest.java

```diff
@@ -0,0 +1,199 @@
+
+        PaytrailRefundCreateRequest.CreateRefundPayload payload = new PaytrailRefundCreateRequest.CreateRefundPayload();
+
+        String refundStamp = UUID.randomUUID().toString();
+        String refundReference = "3759170";
+
+        payload.setRefundStamp(refundStamp);
+        payload.setRefundReference(refundReference);
+        payload.setAmount(10);
+        payload.setEmail("martin.lehtomaa@ambientia.fi");
+
+        PaymentCallbackUrls callbackUrls = new PaymentCallbackUrls();
+
+        callbackUrls.setSuccess("https://ecom.example.com/cart/success");
+        callbackUrls.setCancel("https://ecom.example.com/cart/cancel");
+        payload.setCallbackUrls(callbackUrls);
+
+        PaytrailRefundCreateRequest request = new PaytrailRefundCreateRequest(paymentResponse.getTransactionId(), payload);
+        CompletableFuture<PaytrailRefundCreateResponse> response = client.sendRequest(request);
+
+        IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, () -> refundMapper.to(response.get()));
+        String expectedMessage = "argument \"content\" is null";
+        String actualMessage = exception.getMessage();
+        assertTrue(actualMessage.contains(expectedMessage));
+    }
+
+    @Test
+    public void createTestRefundShopInShop() throws ExecutionException, InterruptedException {
+        PaytrailClient client = new PaytrailClient(aggregateMerchantId, aggregateSecretKey);
+        client.setCustomerMerchantId(shopInShopMerchantId);
+
+        /* Creating a test payment */
+        PaytrailPaymentResponse paymentResponse = createTestShopInShopPayment(client);
+
+        PaytrailRefundCreateRequest.CreateRefundPayload payload = new PaytrailRefundCreateRequest.CreateRefundPayload();
+
+        String refundStamp = UUID.randomUUID().toString();
+        String refundReference = "3759170";
+
+        payload.setRefundStamp(refundStamp);
+        payload.setRefundReference(refundReference);
+        payload.setAmount(10);
+        payload.setEmail("martin.lehtomaa@ambientia.fi");
+
+        RefundItem refundItem = new RefundItem();
+        refundItem.setStamp(UUID.randomUUID().toString());
+        refundItem.setRefundStamp(refundStamp);
+        refundItem.setRefundReference(refundReference);
+        refundItem.setAmount(10);
+        List<RefundItem> items = new ArrayList<>();
+        items.add(refundItem);
+        payload.setItems(items);
+
+        PaymentCallbackUrls callbackUrls = new PaymentCallbackUrls();
+
+        callbackUrls.setSuccess("https://ecom.example.com/cart/success");
+        callbackUrls.setCancel("https://ecom.example.com/cart/cancel");
+        payload.setCallbackUrls(callbackUrls);
+
+        PaytrailRefundCreateRequest request = new PaytrailRefundCreateRequest(paymentResponse.getTransactionId(), payload);
+        CompletableFuture<PaytrailRefundCreateResponse> response = client.sendRequest(request);
```

Comment:
Could you write happy path tests too? I can only see not so happy path tested in here, or do I miss that?

Reply:
Well this will always throw an exception with error text "Transaction not paid", because I did not find a way to programatically make a payment paid. So I do not think there is a way to make a happy path test...

Also the error handling is implemented in a way that the "Transaction not paid" message only gets logged, so we don't get it as return for assertion. I was wondering if I should try to change it so that the error message is returned some how here for assertion. What do you think?